Do not shadow topics with redpanda prefixes

michael-redpanda opened the pull request:
This PR adjusts the auto topic creation strategy.

* Wildcard filter does _not_ select any topic starting with `_redpanda` or `__redpanda`
* Users may specify _literal_ topic names that start with those prefixes if they wish
* Filters will reject any attempt to add literal match to `__consumer_offsets` or `_redpanda.audit_log`

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [X] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
sl: Do not shadow topics with redpanda prefixes

Topics that start with "_redpanda" or "__redpanda" should not be
mirrored unless those topics are specifically included by literal
filters.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Put check in wrong spot

IoannisRP commented on src/v/cluster_link/source_topic_syncer.cc:
```diff
@@ -41,6 +41,9 @@ const absl::flat_hash_set<::model::topic> topic_denylist{
 
+const absl::flat_hash_set<ss::sstring> topic_prefix_denylist{
+  ss::sstring("__redpanda"), ss::sstring("_redpanda")};
+
```
nit: an array/vector could do here as well. the set functionality is not used.

IoannisRP commented on src/v/cluster_link/source_topic_syncer.cc:
```diff
@@ -165,6 +168,19 @@ validate_topic_cache_entry(
       it->second.authorized_operations,
       topic_id);
 }
+
+std::optional<ss::sstring> is_valid_topic(::model::topic_view topic) {
+    if (topic_denylist.contains(topic)) {
+        return ssx::sformat("Topic {} is in the denylist", topic);
+    }
+    for (const auto& prefix : topic_prefix_denylist) {
+        if (topic().starts_with(prefix)) {
+            return ssx::sformat(
+              "Topic {} starts with a denied prefix: {}", topic, prefix);
+        }
+    }
```
nit: `find_if`

IoannisRP reviewed (dismissed):
looks good. Just a couple of nits

michael-redpanda commented:
/ci-repeat 1

bharathv commented on src/v/cluster_link/source_topic_syncer.cc:
```diff
@@ -165,6 +168,19 @@ validate_topic_cache_entry(
+    if (topic_denylist.contains(topic)) {
+        return ssx::sformat("Topic {} is in the denylist", topic);
+    }
+    for (const auto& prefix : topic_prefix_denylist) {
```
this looks a bit brittle.. is it possible the user topics start with this prefix?

```
__redpanda_fastest_queue_in_the_west_twitter_activity_stream
```

I think the idea is to avoid any topics without write_at_offset_stm, so just use the same condition?


```

bool write_at_offset_stm_factory::is_applicable_for(
  const storage::ntp_config& cfg) const {
    return model::is_user_topic(cfg.ntp())
           || cfg.ntp().tp.topic == model::schema_registry_internal_tp.topic;
}
```

If the user doesn't want to replicate these topics, they should set the regex filter accordingly.

michael-redpanda replied:
We will need to document this limitation.  There are also other redpanda "internal" topics that are not internal to the broker but are used by byoc.

michael-redpanda replied:
such as the e2e probe and connectors.  those are regular topics (From the view of RP) but need to be treated as internal ones and not mirrored.

michael-redpanda replied:
How about this - by default `_redpanda` and `__redpanda` prefixed topics are not replicated unless they are specifically included.  And from there I'll have a list of topics that are not allowed to be specifically included (such as the audit log which was the source of these issues).

michael-redpanda committed:
cluster/cl: Adjust rules for patterns

* Reject attempts to have literal matches against the consumer offsets
  or audit log topic
* Do not allow prefix rules for _redpanda or __redpanda named topics

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto: Updated wording for auto create topic filters

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Addressed PR comments

bharathv commented on src/v/cluster_link/source_topic_syncer.cc:
```diff
@@ -39,8 +39,12 @@ const absl::flat_hash_map<ss::sstring, ss::sstring>
 
 const absl::flat_hash_set<::model::topic> topic_denylist{
   ::model::kafka_consumer_offsets_topic,
+  ::model::kafka_audit_logging_topic,
```
I think there is a chance this can go out of sync with `write_at_offset_stm_factory::is_applicable_for()` requirements.. I"m thinking we should factor this out into a common function 
```
model::is_shadow_link_enabled(const topic&);
```

and then use it in both write_at_offset_stm and topic_deny_list here.

That way we are atleast guaranteed that we don't try to DR topics without wao_stm.

michael-redpanda replied:
yeah sounds good I can do that

bharathv reviewed (commented):
lgtm but one refactoring suggestion so wao_stm and topic_denylist don't diverge.

michael-redpanda committed:
model: Add is_shadow_link_enabled function

Function returns true if the selected topic is valid to be used for
shadow linking.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/wao: Use new is_shadow_link_enabled function

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Refactored how valid shadow link topics are known

bharathv commented on src/v/cluster_link/source_topic_syncer.cc:
```diff
@@ -37,9 +38,8 @@ const absl::flat_hash_map<ss::sstring, ss::sstring>
-const absl::flat_hash_set<::model::topic> topic_denylist{
-  ::model::kafka_consumer_offsets_topic,
-};
+const chunked_vector<ss::sstring> topic_prefix_denylist{
```
constexpr std::array<std::string_view... ?

michael-redpanda commented:
Force push:

* Rebased off of dev
